src/bbd/geocoder/geocoder.py

```diff
@@ -0,0 +1,582 @@
+                   [component_formatter(col, self.data.loc[i, col]) 
+                   for col in self.included_cols]}
+
+        #Set default components in defaults
+        for key, value in zip(self.defaults, self.defaults.values()):
+            key, value = component_formatter(key, value)
+            address[key] = value
+
+        result = self._run_geocoder(**address)
+        try:
+            return result.latitude, result.longitude, result.address
+
+        #AttributeError typically occurs if no result from geocoder
+        except AttributeError:
+            #Format Address in proper order.
+            formatted_address = ", ".join([str(address[key]) for key in 
+                                          ['street', 'city', 
+                                          'county', 'state', 
+                                          'country', 'postalcode']
+                                          if key in address])
+            return None, None, formatted_address
+
+
+    def _run_batch(self, batch_list):
+        """Run each item from batch_list through geocoder 
+        saving to disk as it goes and updating self.locations
+        once complete.
+
+        batch - list of indexes from queue to run through 
+                self.geocoder.
+        """
+        #tqdm for progress bar.
+        with open(self.path, "a") as f:
+            desc = f"{self.curr_batch}/{self.tot_batches}"
+            for i in tqdm(batch_list, desc = desc):
+                #For list of str or dataframe of 1 column
+                if (len(self.data.columns) == 1 
+                        or (len(self.included_cols) == 1 
+                            and self.included_cols[0] == 'address')):
+                    lat, lon, address = self._process_address_string(i)
+
+                #For DataFrame or list of dicts
+                else:
+                    lat, lon, address = self._process_address_components(i)
+
+                #Add processed addresses to self.locations
+                self.locations.loc[i, ['latitude', 
+                                       'longitude', 
+                                       'address']] = lat, lon, address
+
+                #Save processed address to path.
+                f.write(f"{i}\t{lat or ''}\t{lon or ''}\t{address}\n")
+
+
+    def run(self, num_batches=None):
+        """
+        """
+        if num_batches is None:
+            num_batches = len(self._queue)//self.batch_size + 1
+            print("Beginning Geocoding: This may take many hours. "\
+                  "Progress saved as it goes.")
+
+        if num_batches <= 0 or len(self._queue) == 0:
+            print("All done!", flush = True)
+            return self.locations
+
+        batch = [] #need to decide how to pull batches from queue
+        for i in range(self.batch_size):
+            try:
+                batch.append(self._queue.popleft())
+            except IndexError:
+                pass #reached end of Queue
```

Comment:
Yep will change to `break`! Yeah I was also debating doing list w/ slices but went with queue just b/c it was one of the last things I added so I went with easy solution. I think the list w/ slices method will lend itself better if run() isn't recursive. 

The other reason I opted for queue was it made it easier to load in and remember its place b/c I don't have to worry about disparate batch_size's between sessions. When I wrote the first version of this (pre-module) I had run() as a loop with list slicing but it was complicated b/c batches would end up redundantly running addresses from previous session if it was loaded again (prob just bad implementation but that was why I choose the easy route).

Reply:
So GeocodeLocations actually picks up where it left off automatically and its actually the driving reason I made the module b/c that happened to me 1 too many times. It is done during the file loading stage which loads the file then trims the queue down to only remaining indices.

For the second question, if they were to concatenate their new data with the old data (with the very important caviot that the index MUST be perserved), they could then just run GeocodeLocations and it would only geocode the data. 

In general I wouldn't recommend doing it that way b/c the easiest way for the results to get messed up is if the indices get reset. Since in order to merge the original data with the new data the dataframes must share an index the safest way to ensure that is to only run GeocodeLocations on a dataframe that was loaded in with no complex manipulations done on it (OR with a well defined user selected index such as VAN ID). 

If someone needs to manipulate their data first, I would recommend doing the manipulations first, then saving the results and then only running GeocodeLocations on the loaded data.

Reply:
And a more concrete answer for the second part. I would have them run the 500 addresses into GeocodeLocations then merge those 500 w/ the results of GeocodeLocations and merge the first 10k with the geocoded results then I would concatenate the 2 merged dataframes. That way there is no room for the geocoded results and the original data being mismatched.

Reply:
Hm. Why bother with the index at all? What if the data set keyed off of the address? That's the part we care about, right? 
`address --> lat, long`
The index is arbitrary. Why doesn't the data structure just make sure that it doesn't duplicate an address, and then we can do away with the index all together?

Reply:
The index is vital b/c it allows the user to merge the result back with the original data w/ the fewest merges. Esp. important because address comes in various forms and will actually be altered when run (trimming off secondary descriptors b.c they break nominatim). Since the use case really calls for connecting the processed data w/ the original data I think we should focus on making that as easy as possible.

Reply:
I agree that it should be easy to join the processed data with the original data -- why not instead add another column for "search address" (that has been trimmed/cleaned) and join on that?